### PR TITLE
[Reservations] Update 2023-04-01 billing swagger with Reservations API paths

### PR DESCRIPTION
**Pull request by gaoyp830**
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [ ] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json [hunk 1]
```diff
@@ -16,20 +16,17 @@
     "application/json"
   ],
   "paths": {
-    "/providers/Microsoft.Billing/billingAccounts/{billingAccountName}/billingProfiles/{billingProfileName}/reservations": {
```
Comment (AdamDotNet) [inline specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json:19]: This path exists in the [previous GA](https://github.com/Azure/azure-rest-api-specs/blob/18e83e24c36eed85a240103b476de0220ace7f84/specification/billing/resource-manager/Microsoft.Billing/stable/2020-05-01/billing.json#L3598).

Are you sure it can be removed? I think ARM team will push back on this.

File: specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json [hunk 2]
```diff
@@ -98,63 +117,197 @@
...
+            "description": "Unexpected error",
+            "schema": {
+              "$ref": "../../../../../common-types/resource-management/v3/types.json#/definitions/ErrorResponse"
+            }
+          }
+        }
+      }
+    },
+    "/providers/Microsoft.Billing/billingAccounts/{billingAccountName}/Microsoft.Capacity/reservationOrders/{reservationOrderId}/reservations": {
+      "get": {
```
Comment (AdamDotNet) [inline specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json:149]: Just curious, does this endpoint support OData filter, search, orderBy, etc.? We've been documenting those parameters in the BillingRP side.

Note that ARM team doesn't want the `$` as part of the path. See [guidelines here](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#query-options).

File: specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json (hunk 1 above)
Comment (AdamDotNet) [reply specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json:19]: Yeah, CI checks are failing

https://github.com/Azure/azure-rest-api-specs/pull/22321/checks?check_run_id=10897962237

> The default tag should contain all APIs. The API path /providers/Microsoft.Billing/billingAccounts/{}/billingProfiles/{}/reservations is not in the default tag. Please make sure the missing API swaggers are in the default tag.

File: specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json [hunk 3]
```diff
@@ -396,6 +751,387 @@
...
+        "Scheduled",
+        "Cancelled"
+      ],
+      "x-ms-enum": {
+        "name": "PaymentStatus",
+        "modelAsString": true
+      }
+    }
+  },
+  "parameters": {
```
Comment (AdamDotNet) [inline specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json:1111]: Some parameters need `"pattern": "<Regex>".

See https://github.com/Azure/azure-rest-api-specs/pull/22321/checks?check_run_id=10897961979

> The resource name parameter 'billingAccountName' should be defined with a 'pattern' restriction.

File: specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/examples/reservationGetByBillingAccount.json [hunk 4]
```diff
@@ -0,0 +1,100 @@
+{
+  "parameters": {
+    "api-version": "2023-04-01",
+    "billingAccountName": "mock_billing_account",
```
Comment (AdamDotNet) [inline specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/examples/reservationGetByBillingAccount.json:4]: In BillingRP, we have the following pattern for billingAccountName:

```json
"billingAccountName": {
  "in": "path",
  "name": "billingAccountName",
  "description": "The ID that uniquely identifies a billing account.",
  "required": true,
  "type": "string",
  "pattern": "^([0-9]+|[0-9A-Fa-f]{8}-([0-9A-Fa-f]{4}-){3}[0-9A-Fa-f]{12}(:[0-9A-Fa-f]{8}-([0-9A-Fa-f]{4}-){3}[0-9A-Fa-f]{12}_[0-9]{4}(-[0-9]{2}){2})?)$",
  "x-ms-parameter-location": "method"
},
```

Once you define patterns for parameters, you samples will have to match the pattern to pass validations.

A modern billing account name that can pass the validation could be the following:

`00000000-0000-0000-0000-000000000000:00000000-0000-0000-0000-000000000000_2019-05-31`

Comment (gaoyp830) [reply specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/examples/reservationGetByBillingAccount.json:4]: This pattern doesnt seem to exist in type.json file where `billingAccountName` is defined. I am going to create the definition separate definition of `billingAccountName` in reservation.json with this pattern

Comment (AdamDotNet) [reply specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/examples/reservationGetByBillingAccount.json]: @gaoyp830 - actually, can you take what types you need from our latest copy of [types.json](https://msazure.visualstudio.com/One/_git/ACE-BillingRP?path=/src/Documentation/swagger2.0/stable/2023-04-01/types.json&version=GBga/2023-04-01&_a=contents) and continue to use reference syntax instead of adding `billingAccountName` within reservations.json?

File: specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json (hunk 2 above)
Comment (gaoyp830) [reply specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/reservation.json:149]: Discussed offline, we should take it to ARM

**Review by AdamDotNet — APPROVED** [review]
We can remove $ in a later PR. Please handle the breaking change process of removing an endpoint as we discussed. Approving for now.

File: specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/examples/reservationsListByBillingAccount.json [hunk 5]
```diff
@@ -12,64 +12,73 @@
...
-                    "grain": 7.0,
+                    "grain": 7,
                     "grainUnit": "days",
-                    "value": 0.05,
+                    "value": 0,
                     "valueUnit": "percentage"
                   },
                   {
-                    "grain": 30.0,
+                    "grain": 30,
```
Comment (chchen3) [inline specification/billing/resource-manager/Microsoft.Billing/stable/2023-04-01/examples/reservationsListByBillingAccount.json:59]: Why is this changed? I see in our code these values are double

**Review by wilcobmsft — APPROVED** [review]
